daq/session_server.py

```diff
@@ -56,12 +69,47 @@ def __init__(self, on_session=None, server_address=None, server_port=None, max_w
 
     def start(self):
         """Start the server"""
-        LOGGER.info('Starting sesson server on ' + self._address)
         self._server.start()
 
     def connect(self, mac, callback):
         """Connect to remote endpoint"""
 
+    def send_device_result(self, mac, device_result):
+        """Connect to remote endpoint"""
+        LOGGER.info('Send device result %s %s', mac, PortBehavior.Behavior.Name(device_result))
+        result = SessionResult(code=SESSION_DEVICE_RESULT[device_result])
+        self._send_reply(mac, SessionProgress(result=result))
+
+    def close_stream(self, mac):
+        """Close the stream"""
+        self._send_reply(mac, False)
+
+    def _send_reply(self, mac, item):
+        self._return_queues[mac].put(item)
+
+    def _init_session(self, request):
+        device_mac = request.device_mac
+        with self._lock:
+            LOGGER.info('New session stream for %s %s/%s',
+                        device_mac, request.device_vlan, request.assigned_vlan)
+            assert device_mac not in self._return_queues, 'already registered %s' % device_mac
+            return_queue = Queue()
+            self._return_queues[device_mac] = return_queue
+        endpoint = SessionEndpoint(ip=('ip-' + device_mac))
+        self._send_reply(device_mac, SessionProgress(endpoint=endpoint))
+        self.send_device_result(device_mac, PortBehavior.Behavior.authenticated)
+        self._on_session(request)
+
+    def _session_stream(self, request):
+        device_mac = request.device_mac
+        while True:
+            item = self._return_queues[device_mac].get()
+            if item is False:
+                break
+            yield item
+        LOGGER.info('Session ended for %s', device_mac)
+        del self._return_queues[device_mac]
+
     def stop(self):
```

Comment:
Should we call close_stream when the session server is stopped?

Reply:
After reading the unit test, it seems that we rely on the module which has set up the server to close the stream. Closing this comment..